src/buffers/disk/leveldb_buffer.rs

```diff
@@ -302,10 +350,22 @@ impl Reader {
             self.delete_offset = new_offset;
 
             let size_deleted = self.unacked_sizes.drain(..num_to_delete).sum();
-            self.current_size.fetch_sub(size_deleted, Ordering::Release);
+            let unread_size = self.current_size.fetch_sub(size_deleted, Ordering::Release);
 
             self.uncompacted_size += size_deleted;
-            if self.uncompacted_size > self.max_uncompacted_size {
+
+            // Compaction can be triggered in two ways:
+            //  1. When size of uncompacted is a percentage of total allowed size.
+            //     Managed by MAX_UNCOMPACTED. This is to limit the size of disk buffer
+            //     under configured max size.
+            let max_trigger = self.uncompacted_size > self.max_uncompacted_size;
+            //  2. When the size of uncompacted buffer is larger than unread buffer.
+            //     Managed partially by MIN_UNCOMPACTED_SIZE. This is to keep the size
+            //     of the disk buffer to be proportional to the unread part of the buffer.
+            let proportional_trigger = self.uncompacted_size > unread_size
+                && self.uncompacted_size >= MIN_UNCOMPACTED_SIZE;
```

Comment:
If I'm understanding this correctly, it seems like most common cases (i.e. reader is keeping up with writer, reasonably large buffer max size) would hit this condition almost exclusively and run a compaction every 1MB. Given that we aim to support at least 10s of MB/sec in throughput, the result would be multiple compactions per second which seems like it could cause performance problems.

Reply:
True, to cover that we can raise the limit to some 40MB. 

Regarding performance impact, from what I saw, there is some fix cost to running compaction but it largely depends on the key interval size. `MIN_UNCOMPACTED_SIZE` exists to manage that balance.